### DISCUSSION TITLE
[FEATURE] Directed graphs

### DISCUSSION THREAD

**Discussion by dr3y** [bokeh/bokeh#11493] [General]
I have been working on making directed graphs in bokeh. Using `networkx` it is easy to create a `DiGraph` object and have directions in my arrows. But using `from_networkx()` there is no way to automatically draw arrows at the end of the graph edges.

My current work-around is to edit the path of `graph.edges` to make it look like an arrow. This has the bad consequence that arrows do not stay arrow-shaped if you resize the graph or zoom in.

![arrow example](https://i.imgur.com/J8doQsj.png)

Proposed solution:
have `from_networkx()` understand about `DiGraph` objects and allow different glyphs to appear at the ends of lines (maybe you want flat head arrows or something instead of normal arrows)

**Comment by IuryPiva** [comment]
Hey @dr3y thanks for submitting this feature request, sorry this took so long to get a response.

Have you looked at our [ArrowHead](https://docs.bokeh.org/en/latest/docs/reference/models/arrow_heads.html?highlight=arrowhead#bokeh.models.arrow_heads.ArrowHead) model? how would your proposal differ from it?

**Reply by dr3y** [reply]
Hi thanks for the reply.

I tried using annotations to make the arrows but they were significantly worse for some reason i don't recall at the moment. Perhaps it was that it was slow to render them or something? Oh also they don't really work when highlighting the lines going into/ out of a node, which is an important feature for me.

**Reply by bryevdv** [reply]
@IuryPiva the issue with arrow heads is that they are annotations,not glyphs, which means they don't show up in legends and don't participate in hit testing. Also it would be reasonable to have automatic conversion from networkx as with undirected graphs. Currently a user would have to manually compute the angle for every arrowhead which is a very high burden (and in case of a dynamic layout possibly impossible at least from Python). I think the old issue is valid but we could do better now by breaking things down into smaller more focused issues

**Reply by mattpap** [reply]
This PR https://github.com/bokeh/bokeh/pull/11122 will effectively add support for directed graphs.

**Reply by sander76** [reply]
I am looking at the above pr, but I still don't have a clue how to use this to add arrows to my directed graph.
Is there some example I could borrow some logic from ?

**Reply by mattpap** [reply]
That PR added experimental support glyph decorations (a few arrow heads) to a handful of glyphs (`Segment` and `Arc`). Complete support for directed graphs will be added as part of NumFocus Small Development Grant that is currently starting. This should be a part of bokeh 3.3 or 3.4. For now you can see [this](https://github.com/bokeh/bokeh/blob/563a487f8155900b170eadc501587c98f154ee15/examples/plotting/trefoil.py) example, to get a preview of how the API will look like.

**Reply by sander76** [reply]
Thanks very much ! I'll dig into it.
